File: forge-core/internal/kerneldecisioncontract/primitives.go
package kerneldecisioncontract

import (
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"unicode"
	"unicode/utf8"

	op "forgeos/forge-core/internal/kerneloperationalcontract"
)

var hashPattern = regexp.MustCompile(`^[0-9a-f]{64}$`)
var identifierPattern = regexp.MustCompile(`^[a-z0-9][a-z0-9._:/-]{0,159}$`)
var legacyAtomPattern = regexp.MustCompile(`^atom-[0-9a-f]{64}$`)
var adrPattern = regexp.MustCompile(`^ADR-[0-9]{4,}$`)

func text(value, label string, maximum int) error {
	if value == "" || len(value) > maximum || !utf8.ValidString(value) {
		return fmt.Errorf("%s must be nonempty UTF-8 text <= %d bytes", label, maximum)
	}
	for _, character := range value {
		if unicode.IsControl(character) || character == 0x061c || character == 0x200e ||
			character == 0x200f || character >= 0x2028 && character <= 0x202e ||
			character >= 0x2066 && character <= 0x2069 {
			return fmt.Errorf("%s contains a forbidden Unicode scalar", label)
		}
	}
	return nil
}

func identifier(value, label string) error {
	if !identifierPattern.MatchString(value) {
		return fmt.Errorf("%s must match identifier grammar", label)
	}
	return nil
}

func hash(value, label string) error {
	if !hashPattern.MatchString(value) {
		return fmt.Errorf("%s must be a lowercase bare SHA-256", label)
	}
	return nil
}

func identity(id, digest, prefix, label string, blank bool) error {
	if blank && id == "" && digest == "" {
		return nil
	}
	if err := hash(digest, label+"_sha256"); err != nil {
		return err
	}
	if id != prefix+digest {
		return fmt.Errorf("%s identity does not bind digest", label)
	}
	return nil
}

func validateAttestations(value DecisionAttestations) error {
	raw, err := canonicalBytes(value, maxStringBytes)
	if err != nil {
		return err
	}
	var values map[string]bool
	if err := json.Unmarshal(raw, &values); err != nil || len(values) != 22 {
		return fmt.Errorf("decision attestations must have exact twenty-two fields")
	}
	for _, enabled := range values {
		if enabled {
			return fmt.Errorf("all twenty-two decision attestations must be false")
		}
	}
	return nil
}

func validatePrincipal(value op.Principal, label string) error {
	if err := text(value.AuthorityDomain, label+".authority_domain", maxShortBytes); err != nil {
		return err
	}
	if err := text(value.PrincipalID, label+".principal_id", maxShortBytes); err != nil {
		return err
	}
	if !oneOf(value.PrincipalType, "agent", "human", "operator", "service") {
		return fmt.Errorf("%s.principal_type is unsupported", label)
	}
	return nil
}

func validateTask(value op.TaskBinding) error {
	strings := []string{value.ChangeID, value.EnvironmentID, value.NodeID,
		value.ProjectID, value.Role, value.RunID, value.TaskID}
	for _, item := range strings {
		if err := text(item, "task_binding field", maxShortBytes); err != nil {
			return err
		}
	}
	if !oneOf(value.EnvironmentClass, "development", "local", "production", "staging", "test") {
		return fmt.Errorf("task_binding.environment_class is unsupported")
	}
	for _, optional := range []*string{value.AttemptID, value.TargetID} {
		if optional != nil {
			if err := text(*optional, "task_binding optional field", maxShortBytes); err != nil {
				return err
			}
		}
	}
	return nil
}

func validateBindings(value op.OperationalBindings) error {
	for _, digest := range []string{value.ContextSHA256, value.EnvironmentSHA256,
		value.PolicySHA256, value.SourceTreeSHA256} {
		if err := hash(digest, "bindings digest"); err != nil {
			return err
		}
	}
	for _, item := range []string{value.EnvironmentProfileID, value.SourceProfileID,
		value.SourceRevision} {
		if err := text(item, "bindings text", maxShortBytes); err != nil {
			return err
		}
	}
	return nil
}

func validateCapability(value op.CapabilityIdentity) error {
	if err := hash(value.CapabilityContractSHA256, "capability_contract_sha256"); err != nil {
		return err
	}
	if err := text(value.CapabilityID, "capability_id", maxShortBytes); err != nil {
		return err
	}
	return text(value.CapabilityVersion, "capability_version", maxShortBytes)
}

func validateStringSet(values []string, label string, maximum int, nonempty bool) error {
	if values == nil || len(values) > maximum || nonempty && len(values) == 0 {
		return fmt.Errorf("%s cardinality is outside frozen bounds", label)
	}
	for _, value := range values {
		if err := identifier(value, label); err != nil {
			return err
		}
	}
	if !sort.StringsAreSorted(values) || duplicateStrings(values) {
		return fmt.Errorf("%s must be strictly sorted and unique", label)
	}
	return nil
}

func duplicateStrings(values []string) bool {
	for index := 1; index < len(values); index++ {
		if values[index-1] == values[index] {
			return true
		}
	}
	return false
}

func oneOf(value string, allowed ...string) bool {
	for _, candidate := range allowed {
		if value == candidate {
			return true
		}
	}
	return false
}

File: forge-runtime/crates/domain/src/kernel_decision_contract/primitives.rs
use std::collections::BTreeSet;

use serde_json::{Map, Value};

use crate::kernel_operational_contract::{
    CapabilityIdentity, OperationalBindings, Principal, TaskBinding,
};

use super::{DecisionAttestations, KernelDecisionContractError, invalid};

pub(super) fn text(
    value: &str,
    label: &str,
    maximum: usize,
) -> Result<(), KernelDecisionContractError> {
    if value.is_empty() || value.len() > maximum {
        return Err(invalid(format!(
            "{label} must be nonempty UTF-8 text <= {maximum} bytes"
        )));
    }
    if value.chars().any(forbidden_scalar) {
        return Err(invalid(format!(
            "{label} contains a forbidden Unicode scalar"
        )));
    }
    Ok(())
}

fn forbidden_scalar(value: char) -> bool {
    value.is_control()
        || matches!(
            value,
            '\u{061c}'
                | '\u{200e}'
                | '\u{200f}'
                | '\u{2028}'..='\u{202e}'
                | '\u{2066}'..='\u{2069}'
        )
}

pub(super) fn identifier(value: &str, label: &str) -> Result<(), KernelDecisionContractError> {
    if value.is_empty() || value.len() > 160 {
        return Err(invalid(format!("{label} is not a frozen identifier")));
    }
    let mut bytes = value.bytes();
    if !bytes
        .next()
        .is_some_and(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit())
        || !bytes.all(|byte| {
            byte.is_ascii_lowercase()
                || byte.is_ascii_digit()
                || matches!(byte, b'.' | b'_' | b':' | b'/' | b'-')
        })
    {
        return Err(invalid(format!("{label} is not a frozen identifier")));
    }
    Ok(())
}

pub(super) fn hash(value: &str, label: &str) -> Result<(), KernelDecisionContractError> {
    if value.len() != 64
        || !value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
    {
        return Err(invalid(format!("{label} must be a lowercase bare SHA-256")));
    }
    Ok(())
}

pub(super) fn identity(
    id: &str,
    digest: &str,
    prefix: &str,
    label: &str,
    allow_blank: bool,
) -> Result<(), KernelDecisionContractError> {
    if allow_blank && id.is_empty() && digest.is_empty() {
        return Ok(());
    }
    hash(digest, &format!("{label}_sha256"))?;
    if id != format!("{prefix}{digest}") {
        return Err(invalid(format!("{label} identity does not bind digest")));
    }
    Ok(())
}

pub(super) fn attestations(
    value: &DecisionAttestations,
) -> Result<(), KernelDecisionContractError> {
    let values = [
        value.approval_authentication_attestation,
        value.authority_attestation,
        value.authorization_attestation,
        value.binding_authentication_attestation,
        value.cas_attestation,
        value.completion_attestation,
        value.content_provenance_attestation,
        value.effect_attestation,
        value.event_append_attestation,
        value.execution_attestation,
        value.grant_authentication_attestation,
        value.hard_guard_attestation,
        value.instruction_attestation,
        value.outcome_attestation,
        value.permission_attestation,
        value.persistence_attestation,
        value.principal_authentication_attestation,
        value.source_resolution_attestation,
        value.transition_attestation,
        value.truth_attestation,
        value.usage_measurement_attestation,
        value.verifier_independence_attestation,
    ];
    if values.iter().any(|item| *item) {
        Err(invalid(
            "all twenty-two decision attestations must be false",
        ))
    } else {
        Ok(())
    }
}

pub(super) fn principal(value: &Principal, label: &str) -> Result<(), KernelDecisionContractError> {
    text(
        &value.authority_domain,
        &format!("{label}.authority_domain"),
        160,
    )?;
    text(&value.principal_id, &format!("{label}.principal_id"), 160)
}

pub(super) fn task(value: &TaskBinding) -> Result<(), KernelDecisionContractError> {
    for item in [
        &value.change_id,
        &value.environment_id,
        &value.node_id,
        &value.project_id,
        &value.role,
        &value.run_id,
        &value.task_id,
    ] {
        text(item, "task_binding field", 160)?;
    }
    for item in [&value.attempt_id, &value.target_id].into_iter().flatten() {
        text(item, "task_binding optional field", 160)?;
    }
    Ok(())
}

pub(super) fn bindings(value: &OperationalBindings) -> Result<(), KernelDecisionContractError> {
    for digest in [
        &value.context_sha256,
        &value.environment_sha256,
        &value.policy_sha256,
        &value.source_tree_sha256,
    ] {
        hash(digest, "bindings digest")?;
    }
    for item in [
        &value.environment_profile_id,
        &value.source_profile_id,
        &value.source_revision,
    ] {
        text(item, "bindings field", 160)?;
    }
    Ok(())
}

pub(super) fn capability(value: &CapabilityIdentity) -> Result<(), KernelDecisionContractError> {
    hash(
        &value.capability_contract_sha256,
        "capability_contract_sha256",
    )?;
    text(&value.capability_id, "capability_id", 160)?;
    text(&value.capability_version, "capability_version", 160)
}

pub(super) fn string_set(
    values: &[String],
    label: &str,
    maximum: usize,
    nonempty: bool,
) -> Result<(), KernelDecisionContractError> {
    if values.len() > maximum || nonempty && values.is_empty() {
        return Err(invalid(format!(
            "{label} cardinality is outside frozen bounds"
        )));
    }
    for value in values {
        identifier(value, label)?;
    }
    if !strict_strings(values) {
        return Err(invalid(format!(
            "{label} must be strictly sorted and unique"
        )));
    }
    Ok(())
}

pub(super) fn strict_strings(values: &[String]) -> bool {
    values
        .windows(2)
        .all(|pair| pair[0].as_bytes() < pair[1].as_bytes())
}

pub(super) fn exact_object<'a>(
    value: &'a Value,
    fields: &[&str],
    label: &str,
) -> Result<&'a Map<String, Value>, KernelDecisionContractError> {
    let object = value
        .as_object()
        .ok_or_else(|| invalid(format!("{label} must be an object")))?;
    let expected = fields.iter().copied().collect::<BTreeSet<_>>();
    let actual = object.keys().map(String::as_str).collect::<BTreeSet<_>>();
    if actual != expected {
        return Err(invalid(format!("{label} fields differ")));
    }
    Ok(object)
}

pub(super) fn object_text<'a>(
    object: &'a Map<String, Value>,
    field: &str,
) -> Result<&'a str, KernelDecisionContractError> {
    object[field]
        .as_str()
        .ok_or_else(|| invalid(format!("{field} must be text")))
}
